### ISSUE TITLE
K8s in Azure breaks internal LB if only one service uses ILB annotation

### ISSUE THREAD

**Issue by UncleTawnos** [Azure/acs-engine#2392]
**Is this a request for help?**:
NO
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
ISSUE
---

**What version of acs-engine?**:
master-latest (cd92d6949c66be875f5708eafb371a04666181fe)
---

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
Kubernetes

**What happened**:
I have deployed fresh cluster (GitVersion:"v1.9.3", GitCommit:"d2835416544f298c919e2ead3be3d0864b52323b) using acs-engine. Major configuration settings:
* RBAC on
* networking - flannel
* custom VNET
* 3 agent pools

I have noticed that if I do any change to a service using ILB annotation `service.beta.kubernetes.io/azure-load-balancer-internal: "true"`, if it's the only one service using ILB, the ILB itself is being deleted and recreated:
```
 "operationName": {
        "value": "Microsoft.Network/loadBalancers/delete",
        "localizedValue": "Microsoft.Network/loadBalancers/delete"
```
```
    "operationName": {
        "value": "Microsoft.Network/loadBalancers/write",
        "localizedValue": "Microsoft.Network/loadBalancers/write"
    },
```

The issue is that, although backend pool is created, no servers are attached to it, rendering everything behind ILB down.

While other ILB-enabled services exists, the operation is just write (with no delete first) as expected 

**What you expected to happen**:
ILB is reconfigured instead of being deleted

**How to reproduce it** (as minimally and precisely as possible):
* Deploy cluster
* create nginx with ILB
* change anything (add a port, label etc)
* observe how ILB is being deleted

**Comment by jackfrancis** [comment]
Can you paste the original api model (scrubbed of creds) that you used to build the cluster? Thanks!

**Comment by UncleTawnos** [comment]
```
{
  "apiVersion": "vlabs",
  "properties": {
    "orchestratorProfile": {
      "orchestratorType": "Kubernetes",
      "orchestratorRelease": "1.9",
      "kubernetesConfig": {
        "networkPolicy": "none",
        "enablePrivateCluster": true,
        "dnsServiceIP": "10.192.0.10",
        "serviceCidr": "10.192.0.0/14",
        "clusterSubnet": "10.244.0.0/16",
        "nonMasqueradeCidr": "10.192.0.0/10"
      }
    },
    "masterProfile": {
      "count": 3,
      "dnsPrefix": "syneacs-prod",
      "vmSize": "Standard_D2_v3",
      "vnetSubnetId": "/subscriptions/00000000-0000-0000-0000-00000000000/resourceGroups/networking/providers/Microsoft.Network/virtualNetworks/k8s-net/subnets/k8s-masters",
      "firstConsecutiveStaticIP": "172.26.1.5",
      "storageProfile": "ManagedDisks",
      "osDiskSizeGB": 127,
      "vnetCidr": "172.26.0.0/16"
    },
    "agentPoolProfiles": [
      {
        "name": "generalpool",
        "count": 4,
        "vmSize": "Standard_D4_v3",
        "availabilityProfile": "AvailabilitySet",
        "storageProfile": "ManagedDisks",
        "osDiskSizeGB": 255,
        "vnetSubnetId": "/subscriptions/00000000-0000-0000-0000-00000000000/resourceGroups/networking/providers/Microsoft.Network/virtualNetworks/k8s-net/subnets/k8s-agents"
      },
      {
        "name": "computepool",
        "count": 4,
        "vmSize": "Standard_F4",
        "availabilityProfile": "AvailabilitySet",
        "storageProfile": "ManagedDisks",
        "osDiskSizeGB": 255,
        "vnetSubnetId": "/subscriptions/00000000-0000-0000-0000-00000000000/resourceGroups/networking/providers/Microsoft.Network/virtualNetworks/k8s-net/subnets/k8s-agents"
      },
      {
        "name": "highmempool",
        "count": 4,
        "vmSize": "Standard_E4_v3",
        "availabilityProfile": "AvailabilitySet",
        "storageProfile": "ManagedDisks",
        "osDiskSizeGB": 255,
        "vnetSubnetId": "/subscriptions/00000000-0000-0000-0000-00000000000/resourceGroups/networking/providers/Microsoft.Network/virtualNetworks/k8s-net/subnets/k8s-agents"
      }
    ],
    "linuxProfile": {
      "adminUsername": "tech",
      "ssh": {
        "publicKeys": [
          {
            "keyData": "ssh-rsa <>"
          }
        ]
      }
    },
    "servicePrincipalProfile": {
      "clientId": "00000000-0000-0000-0000-00000000000",
      "secret": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
    }
  }
}
```

as "nonMasqueradeCidr" results Unknown JSON tag nonMasqueradeCidr I had to "fix" it in azuredeploy.json and azuredeploy.parametes.json manualy

**Comment by jackfrancis** [comment]
@UncleTawnos btw you can configure kubelet `--non-masquerade-cidr` setting explicitly now in your api model. Example usage is:

```
      "kubernetesConfig": {
        ...
        "kubeletConfig": {
            "--non-masquerade-cidr": "10.192.0.0/10"
        }
        ...
      }
```

Also see https://github.com/Azure/acs-engine/blob/master/docs/clusterdefinition.md#kubeletconfig

**Comment by jackfrancis** [comment]
@lachie83 Does the described scenario sound like expected kubernetes behavior? I.e., the ILB is deleted/created when there's only one service behind it, but updates in place when more than one service is behind it.

**Comment by feiskyer** [comment]
> The issue is that, although backend pool is created, no servers are attached to it, rendering everything behind ILB down.

same with https://github.com/Azure/acs-engine/issues/2151#issuecomment-368718505, the fix should be included in next release (v1.9.4)
